File: 529-minesweeper/minesweeper.cpp
class Solution {
public:

    vector<pair<int,int>> getNeighbours(pair<int,int> p, vector<vector<char>>& board, vector<vector<bool>>& vis)
    {
        vector<pair<int,int>> ans;
        vector<pair<int,int>> coords = {{-1,-1},{-1,0},{-1,1},{0,-1},{0,1},{1,-1},{1,0},{1,1}};
        for(int i=0;i<8;i++)
        {
            int x = p.first + coords[i].first;
            int y = p.second + coords[i].second;
            if(x>=0 && y>=0 && x<board.size() && y<board[0].size() && !vis[x][y] && board[x][y]=='E') ans.push_back({x,y});
        }
        return ans;
    }

    int adjacentMines(pair<int,int> p, vector<vector<char>>& board)
    {
        int ans=0;
        vector<pair<int,int>> coords = {{-1,-1},{-1,0},{-1,1},{0,-1},{0,1},{1,-1},{1,0},{1,1}};
        for(int i=0;i<8;i++)
        {
            int x = p.first + coords[i].first;
            int y = p.second + coords[i].second;
            if(x>=0 && y>=0 && x<board.size() && y<board[0].size() && board[x][y]=='M' ) ans++;
        }
        return ans;
    }

    vector<vector<char>> updateBoard(vector<vector<char>>& board, vector<int>& click) {
        queue<pair<int,int>> q;
        vector<vector<bool>> vis(board.size(),vector<bool>(board[0].size(),false));
        if(board[click[0]][click[1]]=='M')
        {
            board[click[0]][click[1]]='X';
            return board;
        }
        q.push({click[0],click[1]});
        vis[click[0]][click[1]]=true;
        while(!q.empty())
        {
            pair<int,int> p = q.front();
            q.pop();
            int count = adjacentMines(p,board);
            if(count>0)
            {
                board[p.first][p.second]='0'+count;
                continue;
            }
            for(auto n : getNeighbours(p,board,vis))
            {
                vis[n.first][n.second]=true;
                q.push(n);
            }
            board[p.first][p.second]='B';
        }
        return board;
    }
};

File: 529-minesweeper/minesweeper.java
class Pair{
    int x;
    int y;

    public Pair(int x,int y){
        this.x = x;
        this.y = y;
    }
}

class Solution {


    public ArrayList<Pair> getNeighbours(char[][] board,int m,int n, Pair p){
        int i = p.x;
        int j = p.y;
        ArrayList<Pair> ans = new ArrayList<>();
        if(i-1>=0 && i-1<m && j>=0 && j<n && board[i-1][j]!='B') ans.add(new Pair(i-1,j));
        if(i-1>=0 && i-1<m && j-1>=0 && j-1<n && board[i-1][j-1]!='B') ans.add(new Pair(i-1,j-1));
        if(i-1>=0 && i-1<m && j+1>=0 && j+1<n && board[i-1][j+1]!='B') ans.add(new Pair(i-1,j+1));

        if(i+1>=0 && i+1<m && j>=0 && j<n && board[i+1][j]!='B') ans.add(new Pair(i+1,j));
        if(i+1>=0 && i+1<m && j-1>=0 && j-1<n && board[i+1][j-1]!='B') ans.add(new Pair(i+1,j-1));
        if(i+1>=0 && i+1<m && j+1>=0 && j+1<n && board[i+1][j+1]!='B') ans.add(new Pair(i+1,j+1));

        if(i>=0 && i<m && j-1>=0 && j-1<n && board[i][j-1]!='B') ans.add(new Pair(i,j-1));
        if(i>=0 && i<m && j+1>=0 && j+1<n && board[i][j+1]!='B') ans.add(new Pair(i,j+1));
        return ans;
    }

    public int adjacentMines(char[][] board,ArrayList<Pair> nodes){
        int count = 0;
        for(Pair p : nodes)
        {
            if(board[p.x][p.y]=='M') count++;
        }
        return count;
    }

    public char[][] updateBoard(char[][] board, int[] click) {
        int x = click[0];
        int y = click[1];
        int m = board.length;
        int n = board[0].length;
        if(board[x][y]=='M')
        {
            board[x][y]='X';
            return board;
        }

        Deque<Pair> dq = new ArrayDeque<>();
        Pair p = new Pair(x,y);
        dq.addLast(p);
        board[x][y] = 'B';

        while(!dq.isEmpty())
        {
            Pair curr = dq.removeFirst();
            ArrayList<Pair> nbs = getNeighbours(board,m,n,curr);
            int nbMines = adjacentMines(board,nbs);
            if(nbMines>0)
            {
                board[curr.x][curr.y]=(char)('0'+nbMines);
                continue;
            }
            else
            {
                for(Pair nb : nbs)
                {
                    board[nb.x][nb.y] = 'B';
                    dq.addLast(nb);
                }
            }

        }
        return board;
    }
}